juju/bundle.py

```diff
@@ -156,311 +171,663 @@ def resolve(self, reference):
+            perform a change.
+        """
         # Fix up values, as necessary.
-        if 'parent_id' in params:
+        params = {}
+        if self.parent_id is not None:
             if params['parent_id'].startswith('$addUnit'):
-                unit = self.resolve(params['parent_id'])[0]
+                unit = context.resolve(params['parent_id'])[0]
                 params['parent_id'] = unit.machine.entity_id
             else:
-                params['parent_id'] = self.resolve(params['parent_id'])
+                params['parent_id'] = context.resolve(params['parent_id'])
 
-        params['constraints'] = parse_constraints(
-            params.get('constraints'))
+        params['constraints'] = parse_constraints(self.constraints)
         params['jobs'] = params.get('jobs', ['JobHostUnits'])
 
-        if params.get('container_type') == 'lxc':
+        if self.container_type == 'lxc':
             log.warning('Juju 2.0 does not support lxc containers. '
                         'Converting containers to lxd.')
             params['container_type'] = 'lxd'
 
         # Submit the request.
         params = client.AddMachineParams(**params)
-        results = await self.client_facade.AddMachines(params=[params])
+        results = await context.client_facade.AddMachines(params=[params])
         error = results.machines[0].error
         if error:
             raise ValueError("Error adding machine: %s" % error.message)
         machine = results.machines[0].machine
         log.debug('Added new machine %s', machine)
         return machine
 
-    async def addRelation(self, change_id, endpoint1, endpoint2):
-        """
-        :param endpoint1 string:
-        :param endpoint2 string:
-            Endpoint1 and Endpoint2 hold relation endpoints in the
-            "application:interface" form, where the application is always a
-            placeholder pointing to an application change, and the interface is
-            optional. Examples are "$deploy-42:web" or just "$deploy-42".
-        """
-        ep1 = self.resolveRelation(endpoint1)
-        ep2 = self.resolveRelation(endpoint2)
+    def __str__(self):
+        machine = "new machine"
+        if self.container_type is not None and self.container_type != "":
+            machine = "{container_type} container on {machine}".format(container_type=self.container_type,
+                                                                       machine=machine)
+        return "add {}".format(machine)
 
-        log.info('Relating %s <-> %s', ep1, ep2)
-        return await self.model.add_relation(ep1, ep2)
 
-    async def deploy(self, change_id, charm, series, application, options, constraints,
-                     storage, endpoint_bindings, *args):
-        """
-        :param charm string:
-            Charm holds the URL of the charm to be used to deploy this
-            application.
+class AddRelationChange(ChangeInfo):
+    _toPy = {'endpoint1': 'endpoint1',
```

Comment:
why do we need _toPy here as well?

Reply:
Before closing out this PR, I wanted to address this comment. So the reason this exists is for the following reasons (although they might not be good enough reasons?):

 1. Conversion between what juju outputs and what pylib expects is somewhat different and the meshing of the different languages to types is somewhat problematic. Luckily it's a well trodden path in pylib and this is the same convention that's in the client facade code. So if you're hacking away at pylib it should be rather familiar to use this.
 2. We want an exact match of expected values, you don't get anything for free, so you are forced to add to the values. This has previously bitten us before, where we get some new values, unsure how to deal with them, so we either drop them or pass them on directly, either way ends up being wrong. So with this, we're saying what we want to pluck from the api response, anything else is assigned None.
 3. All values that aren't found in the api response are assigned None directly, without having to check if it exists or not. It removes a lot of if, else code.

I'm not sold on the name `_toPy` and we could change it to make more sense, but that's currently in the rest of the code base, so reuse and familiarity might be worth something?